File: Palindrome_Partitioning_131.java
//Time Complexity : O(m^n)
//Space Complexity : O(n)
//Did this code successfully run on Leetcode : Yes
//Any problem you faced while coding this : None

import java.util.List;
import java.util.ArrayList;

class Palindrome_Partitioning_131 {
    List<List<String>> result = new ArrayList<>();
    public List<List<String>> partition(String s) {
        if(s.length() == 0)
            return result;
        helper(s, new ArrayList<>(), 0);
        return result;
    }
    
    private void helper(String str, List<String> temp, int index)
    {
        // Base
        if(index == str.length())
        {
            result.add(new ArrayList<>(temp));
            return;
        }
        // Logic
        for(int i = index; i < str.length(); i++)
        {
            if(isPalindrome(str, index, i))
            {
                // Action
                temp.add(str.substring(index, i+1));
                // Recurse
                helper(str, temp, i+1);
                // Backtrack
                temp.remove(temp.size() - 1);
            }
        }
    }
    
    private boolean isPalindrome(String str, int left, int right)
    {
        if(left == right)
            return true;
        while(left < right)
        {
            if(str.charAt(left) != str.charAt(right))
                return false;
            left++;
            right--;
        }
        return true;
    }
}

File: PalindromE_Partitioning131.py
class Solution:
    def partition(self, s: str) -> List[List[str]]:
        # Note don't change sequence
        # Time complexity = O(N*(2**N))
        # Space complexity = O(N)
        result = []
        if s == None or len(s) == 0:
            return result

        def helper(s, pivot_index, path):
            # base
            if pivot_index >= len(s):
                result.append(path.copy())
                return

            # logic
            for i in range(pivot_index, len(s)):
                substr = s[pivot_index: i + 1]
                if isPalindrome(substr):
                    # action
                    path.append(substr)
                    # recurse
                    helper(s, i + 1, path)
                    # backtrack
                    path.pop()

        def isPalindrome(s):
            start = 0
            end = len(s) - 1
            while start < end:
                if s[start] != s[end]:
                    return False
                start += 1
                end -= 1
            return True

        helper(s, 0, [])
        return result